### ISSUE TITLE
No staging area upload "strategy"

### ISSUE THREAD

**Issue by yarikoptic** [dandi/dandi-api#116]
#104 is our current approach to orchestrate upload+validation through a staging area.  Cons of it will be: transferring key from staging into the key store would take awhile for any sizeable file (since there is no "rename" or "mv" within AWS S3 AFAIK) which could potentially lead to various side-effects and just waste of "resources".  In a brief chat with @satra last week I have voiced an idea to (later) RF to use another approach which should avoid such "expensive" transfer from staging area.  I am yet to look at current implementation to provide a bit more "grounded" explanation or prototypical code but it boils down to:

- upon initiation of upload, lock a blob (given sha256) for that new upload id.  If some other process is trying to upload the same blob at the same time -- return 503 Service Unavailable with `Retry-After` with some reasonable amount set (e.g. `min(ETA, 5 minutes)`)
- upon receiving the lock, we initiate (multipart) upload to the target location (not staging area)
- to avoid stale uploads which would forbid re-upload, possible solutions
  - if possible (in particular for multi part uploads) we can monitor (once in a while) if a known upload job target blob size increased from the time of previous check.  That would also allow to estimate ETA to be possibly reported in `/uploads` and aforementioned `Retry-After`
  - client might need to be instructed to send a "ping" for every ongoing upload (may be with its idea on how much was already uploaded to assist in ETA estimation) once in a while (e.g. ~1 minute).  If there are uploads without a ping within the past 5 minutes, we delete those uploads.
- upon completion of the upload and successful validation of checksum - lock is removed, and blob is added to blobs DB (before that it is known to exist only within uploads)
- deletion of incomplete or invalidated upload (checksum mismatch only, metadata must be validated upon initialization of the upload) would delete partially uploaded key and remove the lock.  Relates to #115 (which should be there for any implementation IMHO)

I think this might work quite well:
- sha256 collision should happen only for tiny/small files, so Retry-After request should be sufficient to just wait and likely just get a response that the sha256 is already in the blob store upon the next attempt
- for any actual data file, there must be no collision really: it is unlikely that two researchers would upload exactly the same asset in parallel for the same or different dandisets.  If so happens -- most likely it is the same uploader, trying to upload identical files under different paths or alike.  Would just increase efficiency since then would upload just once ;-)
- Having #115 implemented would be more important for such workflow (than e.g. for staging area one) to make it possible for the client to be interrupted and then try to re-upload the same file right away.

**Comment by satra** [comment]
@yarikoptic - there is also a brief synopsis in this issue which was closed (https://github.com/dandi/dandi-api/issues/103#issuecomment-777620133)

**Comment by yarikoptic** [comment]
Oh, right! Sorry I have missed the ping. Let's keep this issue open until implemented or closed as won't do ;-)

**Comment by yarikoptic** [comment]
it was implemented but PR did not close this -- we upload now into a unique blob_id (uuid) which is different from digest/etag.
